### PR TITLE
[Do not merge] Add RepoData.json and consumes api output

### PR DESCRIPTION
**Pull request by chcosta**
I've generated RepoData.json for CoreFx.  Below is the RepoData.json file I created, and the output in the form of a consumes.json file.

Does any of this look unexpected?  I don't intend to check this in until RepoUtil is available in CoreFx.

/cc @weshaggard @dagood @jhendrixMSFT @markwilkie 

I have a simple (read 2-sentence) design doc [here](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/consumes-api.md)

It seems that there is redundancy between the data I'm including in RepoData.json and that which @ericstj has catalogued in https://github.com/dotnet/corefx/blob/master/pkg/Microsoft.Private.PackageBaseline/packageIndex.json.  We need to consider some way to consolidate this.

### CODE REVIEW

**Comment by chcosta** [comment]
I'm not certain how to handle test-runtime.  I don't think the RepoUtil has support for a project dependency include in a project.json.

File: RepoData.json [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        "Microsoft.Portable.Targets": "0.1.1-dev",
+        "Microsoft.TargetingPack.Private.NETNative": "1.0.1-beta-24430-00",
+        "Microsoft.TargetingPack.Private.WinRT": "1.0.1",
+        "Microsoft.Win32.Primitives": [ "4.0.0" ],
+        "Microsoft.Win32.Registry": [ "4.0.0" ],
+        "Microsoft.xunit.netcore.extensions": "1.0.0-prerelease-00704-03",
+        "Newtonsoft.Json": "8.0.4-beta1",
+        "OpenCover": "4.6.519",
+        "ReportGenerator": "2.4.3",
+        "System.AppContext": [ "4.0.0" ],
```
Comment (dagood) [inline RepoData.json]: Why are some of these arrays and some aren't? I think tooling should treat singleton arrays the same as a string, but I'm curious why in these cases.

File: consumes.json [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+    "System.Xml.ReaderWriter": "4.3.0-beta-24430-01",
+    "System.Xml.XDocument": "4.3.0-beta-24430-01",
+    "System.Xml.XmlDocument": "4.3.0-beta-24430-01",
+    "System.Xml.XmlSerializer": "4.3.0-beta-24430-01",
+    "System.Xml.XPath": "4.3.0-beta-24430-01"
+  },
+  "toolset": {
+    "Microsoft.DotNet.BuildTools": "1.0.26-prerelease-00726-03",
+    "Microsoft.DotNet.BuildTools.RepoUtils": "1.0.0-prerelease-00726-03",
+    "test-runtime": null
```
Comment (dagood) [inline consumes.json]: I think we can safely exclude `"target": "project"` dependencies like test-runtime since they aren't `consume`d as packages.

**Comment by dagood** [comment]
I'm wondering about those stable and prerelease versions... it's true that the build does consume them, but with a major purpose of the repo API being to make it easier to build from source, does it make sense for `consumes` to list them? I could see them being "build-time tools", but that might be a bit of a stretch. Does the repo API need to consider build type?
/cc @ellismg @jaredpar

File: RepoData.json (hunk 1 above)
Comment (chcosta) [inline RepoData.json]: I will update the singleton arrays to be strings, that was an artefact of the tooling I wrote which generated the entries then stripped out the prerelease versions.

File: RepoData.json [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        "System.Xml.XmlSerializer": "4.0.0",
+        "System.Xml.XPath": "4.0.0",
+        "xunit": "2.1.0",
+        "xunit.console.netcore": "1.0.3-prerelease-00607-01",
+        "xunit.runner.console": "2.1.0",
+        "xunit.runner.utility": "2.1.0"        
+    },
+
+    "toolsetPackages": [
+        "Microsoft.DotNet.BuildTools",
```
Comment (weshaggard) [inline RepoData.json]: How does this relate, if at all, to the other tool versions we have been talking about?

Comment (chcosta) [inline RepoData.json]: Build Dependencies: artifacts which are referenced by the build output and include NuGet packages, MSI, etc ... In order to support a composed build these are further divided into floating and static dependencies: ◦Floating: dependencies where versions can change as a part of a composed build via the  change  command. This commonly includes CoreFx, CoreClr, etc ...

◦Static: dependencies whose version do not change via the  change  command. This is commonly used for SDK binaries, legacy tools, etc ...


•Toolset Dependencies: artifacts used in the production of build output. This commonly includes NuGet.exe, compilers, etc ...

**Comment by chcosta** [comment]
Regarding the redundancy between RepoData "fixedPackages" and CoreFx's packageIndex.json, I'd like to do two things:

1. Break the "StableVersions" out of packageIndex.json into its own separate json file
 - In https://github.com/dotnet/corefx/commit/9cd6cbc2edaea6e822ce86c9e8ee0ae29b6f84ce, @ericstj collapsed the stable versions, baselines, assembly versions, etc.. into one file.  I'd like to undo part of that but as json instead of msbuild props files.
 - I'll need to update BuildTools to understand both a packageIndex.json and a stableVersions.json
2. Add a "stableVersions" section to RepoData.json which also allows an "import".  "stableVersions" would be a set of library / versions which aren't part of consumes "fixed" api output unless those libraries / versions are detected in a project.json. @jaredpar

**Comment by ericstj** [comment]
@chcosta can you elaborate on why you need StableVersions in repo data?  In packaging we need this for dependency harvesting and package dependency generation.  It's meant to be a DB of all known packages in the past and present.  If I understand correctly Repo data is meant to be only a representation of the present.

**Comment by chcosta** [comment]
Many (not all) of the versions in StableVersions are representative of the present in our repo.  I'd like to use a combination of stable versions and our current project.json to determine which package references are fixed packages.  For example, src\Microsoft.CSharp\ref\project.json references the System.Runtime package version 4.0.0 which is a stable version.

**Comment by ericstj** [comment]
Any reason why you need a separate data file for that?  Why not load the PackageIndex and use it?  It sounds like your scenario is for a subset of what is already supported.

**Comment by chcosta** [comment]
Yes, my scenario is for a subset.  The packageIndex concept doesn't apply to every repo which will provide produces / consumes api support.  I don't want to teach the produces / consumes api tools about packageIndex because it would require unnecessary overhead.  I think that teaching the tool about stable versions is a good compromise.  I also find this separation preferable over moving the subset I care about into repodata and having to consume that into CoreFx.

**Comment by chcosta** [comment]
I had a chat with @ericstj regarding the intent of packageIndex and its use in this scenario.  While, there's not huge push back to the concept, we both agreed that the purpose of RepoData.json is not to necessarily catalogue things which we are able to dynamically determine from the state of the repo, but instead to provide overrides to those things which are non-standard to the repo.  

Another direction to tackle this, without requiring a tight coupling between RepoData.json and PackageIndex.json, is to dynamically determine stable versions, ie permit an "InferStableVersions:true" concept.  These stable dependencies would automatically become fixed packages in the consumes api output.  If there are fixed dependencies which we want to make floating dependencies (the opposite of any concept currently in RepoUtil), then we would need to permit an override for that scenario.  

The concern with dynamically making stable dependencies into fixed packages is when, for example, we ship a stable version and update our test tree to depend on it.  I'd imagine that this would then break any update logic which @dagood is implementing based on the consumes api output.  Perhaps that is a unique enough scenario that we should manually update those dependencies anyways (that's what we do today, I believe).

**Comment by chcosta** [comment]
The latest version of the files in this PR show the RepoData.json file I would commit into CoreFx and the corresponding output (consumes.json), which would be produced using the RepoUtil tool (`"RepoUtil.exe consumes"`) with stable dependency versions automatically converted to fixed packages.

**Comment by dagood** [comment]
So it sounds like the plan now is to use RepoData.json, and draw info from the project.jsons. In earlier discussions we weren't planning on using the project.jsons as the source of truth--has this changed? I'm fine if it has, it's just a significant change in direction.

Reading about the PackageIndex, I see my earlier plan to check in the raw `consumes` file would duplicate it. I'm still wondering [if/when those should be included in the output](https://github.com/dotnet/corefx/pull/11350#issuecomment-244226110), though. Maybe there's a better place to discuss this?

**Comment by chcosta** [comment]
> In earlier discussions we weren't planning on using the project.jsons as the source of truth--has this changed?

Where can I find out more about the earlier plan?  What was the source for this data intended to be?

**Comment by dagood** [comment]
The email thread "Adding Produces / Consumes API's to CoreFx" is what I'm thinking of, and I expanded on it some in the "Dependency auto-update" meeting. It hasn't been communicated very concretely yet--my main concern is just making sure I haven't missed anything. :smile:

I'm going to put up a PR for a doc in buildtools (later today is the plan), but my current thoughts conflict with this PR. I think that stable dependencies question is also significant in how this ends up. If it's better to get this PR in as-is I could just write it and we can discuss differences over there.

**Comment by chcosta** [comment]
Let's sync up offline when you get a chance.  My read of the "Adding Produces / Consumes API's to CoreFx" is that it is very closely tied to project.jsons.  Perhaps discussing what repodata.json represents and understanding your current thoughts would help us align.

**Comment by stephentoub** [comment]
@chcosta, is this still in progress?

**Comment by chcosta** [comment]
@stephentoub Yes, this is still in progress.  Sorry for the delay, release/1.0.0 servicing work has taken priority.

**Comment by karelz** [comment]
@chcosta do you plan to finish it this week? If not, can we please close it for now and reopen when you have time to work on it? (Trying to cleanup "stale" PRs)
Thanks!

**Pull request closed without merging**